### ISSUE TITLE
Functionals over complex spaces

### ISSUE THREAD

**Issue by adler-j** [odlgroup/odl#590]
For some applications (notably MRI, but possibly also phase contrast) it would be interesting to look at data spaces that are complex. In that case, we need to support this for the functionals as well.

##### Convex conjugate

To my knowledge, the issue with this is mostly that we cannot define the convex conjugate properly. The definition is usually:

```
f^*(x^*) = sup_x { <x^*, x> - f(x) }
```

This obviously becomes problematic in case `X` is complex, since then `<x^*, x>` is complex valued and this doesn't make any sense. One way to solve this, used in [this article](http://download.springer.com/static/pdf/491/art%253A10.1007%252Fs001860050004.pdf?originUrl=http%3A%2F%2Flink.springer.com%2Farticle%2F10.1007%2Fs001860050004&token2=exp=1474387469~acl=%2Fstatic%2Fpdf%2F491%2Fart%25253A10.1007%25252Fs001860050004.pdf%3ForiginUrl%3Dhttp%253A%252F%252Flink.springer.com%252Farticle%252F10.1007%252Fs001860050004*~hmac=be918533964713cf8a2eeaa48a299b9d16ab906b51a252af327b85f7aebacdc6) is to simply take the real part:

```
f^*(x^*) = sup_x { Re <x^*, x> - f(x) }
```

At least for the `|| . ||_2^2` functional, this gives the usual result, and I'm quite sure we get nice results for the norms as well.

**Comment by adler-j** [comment]
Bumping this. I'll go ahead with the #587 PR merge, but it causes the `douglas_rachford_primal_dual_mri` example to fail, since we need to define a L2 functional with a complex domain.

We could use @aringh input here.

**Comment by adler-j** [comment]
Bumping this issue again, we really need to fix this since this change introduced problems down the line.

**Comment by kohr-h** [comment]
Yes. Mathematically I think we should go for the "real part" solution since it effectively identifies C with R^2, which is probably what you usually want. In many cases we can remap the convex conjugate to well-known ones for real spaces, for example the 1-norm would correspond to the group norm 1,2 (or 2,1?).
